cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -1803,37 +1805,63 @@ class AzureFileSystem::Impl {
                            const AzureLocation& location, bool recursive) {
     DCHECK(!location.container.empty());
     DCHECK(!location.path.empty());
-    // Non-recursive CreateDir calls require the parent directory to exist.
-    if (!recursive) {
+    if (recursive) {
+      std::vector<AzureLocation> target_locations;
+      // Recursive CreateDir calls require there is no file in parents.
+      auto parent = location;
+      while (!parent.path.empty()) {
+        ARROW_ASSIGN_OR_RAISE(auto info, GetFileInfo(container_client, parent));
+        if (info.type() == FileType::File) {
+          return NotADir(parent);
+        }
+        if (info.type() == FileType::NotFound) {
+          target_locations.push_back(parent);
+        }
+        parent = parent.parent();
```

Comment:
A better name for the `parent` variable is `prefix` because a "prefix" can be the entire location whereas starting with `parent = location;` is very weird because of the choice of variable name.

Reply:
OK. I've changed it to `prefix`.